app/commands/track/determine_concepts_layout.rb

```diff
@@ -0,0 +1,142 @@
+
+# ConceptGraph transforms a track configuration to a form compatible for manipulating for
+# displaying the progression of the track
+class Track
+  class DetermineConceptsLayout
+    include Mandate
+
+    def initialize(track_config)
+      @graph = Graph.new(track_config['exercises'])
+    end
+
+    # Layout computes the ordering of the exercises in a track by creating
+    # a graph then traversing it to discover and memoize the progression
+    def call
+      {
+        levels: levels,
+        connections: connections
+      }
+    end
+
+    private
+    attr_reader :graph
+
+    memoize
+    def levels
+      return [] if graph.empty?
+
+      raise TrackHasCyclicPrerequisiteError if graph.has_cycle?
+
+      Array.new(graph.levels.max + 1) { [] }.tap do |level|
+        graph.levels.each_with_index do |level_idx, node_idx|
+          node = graph.node_for_index(node_idx)
+          node.level = level_idx
+
+          level[level_idx] << node
+        end
+      end
+    end
+
+    memoize
+    def connections
+      levels.drop(1).with_index.flat_map do |(level, level_idx)|
+        level.flat_map do |node|
+          node.prerequisites.
+            map { |prerequisite| graph.node_for_concept(prerequisite) }.
+            reject(&:nil?).
+            uniq.
+            select { |prerequisite_node| prerequisite_node.level + 1 == level_idx }.
+            map { |prerequisite_node| { from: prerequisite_node.slug, to: node.slug } }
+        end
+      end
+    end
+
+    class Graph
+      include Mandate
+
+      # Edge representing a directed edge
+      Edge = Struct.new(:from, :to, keyword_init: true)
+
+      # Node for representing an exercise within a track
+      Node = Struct.new(:index, :slug, :uuid, :concepts, :prerequisites, :level, keyword_init: true)
+
+      def initialize(exercises)
+        @exercises = exercises
+        @nodes = determine_nodes
+        @edges = determine_edges
+      end
+
+      memoize
+      def empty?
+        nodes.empty?
+      end
+
+      memoize
+      def has_cycle?
+        GraphUtils.contains_cycle?(adjacencies)
+      end
+
+      def node_for_concept(concept)
+        nodes_keyed_by_concept[concept]
+      end
+
+      def node_for_index(index)
+        nodes_keyed_by_index[index]
+      end
+
+      memoize
+      def levels
+        GraphUtils.calculate_levels(adjacencies)
+      end
+
+      private
+      attr_reader :nodes, :edges, :exercises
+
+      # Creates adjacency list for a graph with directed edges
+      memoize
+      def adjacencies
+        Array.new(nodes.length) { [] }.tap do |adjacencies|
+          edges.each do |edge|
+            from = edge.from.index
+            to = edge.to.index
+            adjacencies[from] << to
+          end
+        end
+      end
+
+      def determine_nodes
+        exercises.map.with_index do |exercise, idx|
+          Node.new(
+            index: idx,
+            slug: exercise['slug'],
+            uuid: exercise['uuid'],
```

Comment:
Two thoughts:
1. I think it would be worth passing in the Concept object (which you'll get from the code above: https://github.com/exercism/v3-website/pull/112/files#r499166532). We can then use that to generate URLs, get their names, etc.
2. We don't use this uuid field anywhere as far as I can see.